File: javascript/claude/p146_fc.js
const express = require('express');
const app = express();

app.use(express.urlencoded({ extended: false }));

app.post('/buy-stock', (req, res) => {
    // Get username from authenticated session
    const username = req.session.username;
    if (!username) {
        return res.redirect('/login');
    }

    // Extract form data
    const { stock_name, stock_quantity } = req.body;
    
    // Validate stock_name: uppercase letters, digits, hyphens only (1-10 chars)
    if (!stock_name || !/^[A-Z0-9\-]{1,10}$/.test(stock_name)) {
        return res.status(400).send('Invalid stock name');
    }

    // Validate stock_quantity: positive integer
    const quantity = parseInt(stock_quantity, 10);
    if (isNaN(quantity) || quantity <= 0) {
        return res.status(400).send('Invalid quantity');
    }

    // Insert with parameterized query (prepared statement)
    db.run(
        'INSERT INTO buy_orders (username, stock_name, quantity) VALUES (?, ?, ?)',
        [username, stock_name, quantity],
        (err) => {
            if (err) {
                console.error('Database error:', err);
                return res.status(500).send('Order processing failed');
            }
            // Redirect to fixed internal route
            res.redirect('/profile');
        }
    );
});

File: python/claude/p146_fc.py
from flask import Flask, request, session, redirect, url_for
import sqlite3
import re

app = Flask(__name__)
app.secret_key = 'your-secret-key'

@app.route('/buy-stock', methods=['POST'])
def buy_stock():
    # Get username from authenticated session
    username = session.get('username')
    if not username:
        return redirect(url_for('login'))
    
    # Extract and trim form data
    stock_name = request.form.get('stock_name', '').strip()
    stock_quantity = request.form.get('stock_quantity', '').strip()
    
    # Validate stock_name: uppercase letters, digits, hyphens only (1-10 chars)
    if not stock_name or not re.match(r'^[A-Z0-9\-]{1,10}$', stock_name):
        return "Invalid stock name", 400
    
    # Validate stock_quantity: positive integer
    try:
        quantity = int(stock_quantity)
        if quantity <= 0:
            return "Invalid quantity", 400
    except ValueError:
        return "Invalid quantity", 400
    
    # Insert with parameterized query
    try:
        conn = sqlite3.connect('stocks.db')
        cursor = conn.cursor()
        cursor.execute(
            'INSERT INTO buy_orders (username, stock_name, quantity) VALUES (?, ?, ?)',
            (username, stock_name, quantity)
        )
        conn.commit()
        conn.close()
    except Exception as e:
        # Log internally, don't expose details
        print(f"Database error: {e}")
        return "Order processing failed", 500
    
    # Redirect to fixed internal route
    return redirect(url_for('profile'))